include/fuzzuf/optimizer/havoc_optimizer.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#ifndef FUZZUF_INCLUDE_OPTIMIZER_HAVOC_OPTIMIZER_HPP
+#define FUZZUF_INCLUDE_OPTIMIZER_HAVOC_OPTIMIZER_HPP
+
+#include "fuzzuf/optimizer/optimizer.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::optimizer {
+
+/**
+ * @class HavocOptimizer
+ * @brief Optimizer for two parameters used in `Mutator::Havoc`
+ * @details This class will work as a base class in `Mutator::Havoc`.
+ * It has two virtual methods:
+ *  - `CalcBatchSize` should return the number of times mutation operators
+ *    are applied to a single input.
+ *  - `CalcMutop` should return an index of mutation operator that
+ *     will be used for the `batch_idx`-th time out of `CalcBatchSize()` times.
+ * If you want to implement a new algorithm that controls the two parameter,
+ * you should define a new class that derives from this class.
+ **/
+class HavocOptimizer {
+ public:
+  HavocOptimizer() {}
+  virtual ~HavocOptimizer() {}
+
+  virtual u32 CalcBatchSize() = 0;
+  virtual u32 CalcMutop(u32 batch_idx) = 0;
+};
```

Comment:
Currently, results of the previous run are feedbacked in `CalcBatchSize` implicitly. This implementation assumes that the information such as `IsMutopBanned` are available when `CalcBatchSize` is executed, which is not ideal. It is also not ideal for `CalcBatchSize` to include an feedback process. This can be solved by separating the functions like follows:
```cpp
class HavocOptimizer {
 public:
  HavocOptimizer() {}
  virtual ~HavocOptimizer() {}
  u32 UpdateAndGetSize() {
    UpdateInternalState();
    return GetBatchSize();
  }
  virtual u32 GetMutop(u32 batch_idx) = 0;

private:
  virtual u32 GetBatchSize() = 0;
  virtual void UpdateInternalState() = 0;
};
```
This makes it clear from the naming that the internal state can be updated here, and also provides the separated process. This change will affect all Optimisers and will require some work there.

Reply:
Fixed in https://github.com/fuzzuf/fuzzuf/pull/92/commits/bb6dd296fe1805275d2bbe9d5b1be61bf47d0d82